### PR TITLE
Keep track of recently rejected transactions.

### PR DESCRIPTION
**Pull request by dajohi**
This prevents the node from repeatedly requesting and rejecting the
same transaction as different peers inv the same transaction.

Idea from Bitcoin Core commit 0847d9cb5fcd2fdd5a21bde699944d966cf5add9.

[Review on Reviewable](https://reviewable.io/reviews/btcsuite/btcd/484)

### CODE REVIEW

File: blockmanager.go [hunk 1]
```diff
@@ -1069,6 +1106,40 @@ func (b *blockManager) handleInvMsg(imsg *invMsg) {
 	}
 }
 
+// limitNumRejectedTxns limits the number of rejected transactions by evicting
+// a random rejected transaction if adding a new one would cause it to overflow
+// the max allowed.
+func (b *blockManager) limitNumRejectedTxns() error {
```
Comment (davecgh) [inline blockmanager.go]: I'd prefer this be defined above the only function it's called from so it's consistent with the vast majority of the code base.

**Comment by davecgh** [comment]
Other than my comment this looks good.  I'll do some manual testing on it as well.

**Comment by dajohi** [comment]
I will wait on this as I'd like to add limits to requestedTxns and requestedBlocks as well.

**Comment by dajohi** [comment]
I wonder if this should wait til after the #445?

File: blockmanager.go [hunk 2]
```diff
@@ -479,11 +496,20 @@ func (b *blockManager) handleTxMsg(tmsg *txMsg) {
 	// already knows about it and as such we shouldn't have any more
 	// instances of trying to fetch it, or we failed to insert and thus
 	// we'll retry next time we get an inv.
-	txHash := tmsg.tx.Sha()
 	delete(tmsg.peer.requestedTxns, *txHash)
 	delete(b.requestedTxns, *txHash)
 
 	if err != nil {
+		// Do not request this transactions again until a new block
```
Comment (wallclockbuilder) [inline blockmanager.go]: We're working with one tx here. Update comment to reflect. ~~transactions~~ transaction.

**Comment by davecgh** [comment]
@dajohi: I know this is waiting on #445 to go in, but would you please address the comments and rebase it on top of 445.  I suspect that one should be going in within a week or so and I'd like to have this one ready.  Thanks!

**Comment by davecgh** [comment]
#445 has been merged.  Please update this accordingly.

**Comment by dajohi** [comment]
rebased.

File: blockmanager.go [hunk 3]
```diff
@@ -1072,6 +1141,40 @@ func (b *blockManager) handleInvMsg(imsg *invMsg) {
...
 }
 
+// limitMap is a helper function for maps that require a maximum limit by
+// evicting a random rejected transaction if adding a new value would cause it
+// to overflow the maximum allowed.
+func (b *blockManager) limitMap(m map[wire.ShaHash]struct{}, limit int) error {
+	if len(m)+1 > limit {
+		// Generate a cryptographically random hash.
+		randHashBytes := make([]byte, wire.HashSize)
+		_, err := rand.Read(randHashBytes)
```
Comment (Roasbeef) [inline blockmanager.go]: So, I'm not sure we really need strong _cryptographic_ randomness here. I suspect reading from `/dev/urandom` may become an unnecessary bottleneck when processing transactions. In order to speed up the execution of this function, we may want to instead use a fast PRNG such as [xorshift](https://en.wikipedia.org/wiki/Xorshift). There exists a pure-go implementation, which can be found [here](https://github.com/lazybeaver/xorshift). 

NOTE: I haven't benchmarked this yet, but this is a recommendation based on past experience when optimizing the `SigCache` in #598.

Comment (davecgh) [inline blockmanager.go]: Yeah, you have a fair point and I think that's probably a reasonable place to look at for optimizing.

However, given this is really just a refactor of the existing code, I don't suspect it should hold up this PR.

**Pull request merged**